### PR TITLE
[ROCm] Enable XLA python callback on ROCm.

### PR DESCRIPTION
**Pull request by rsanthanam-amd**
/cc @hawkinsp

### CODE REVIEW

**Comment by rsanthanam-amd** [comment]
Corresponding PR in TF:

https://github.com/tensorflow/tensorflow/pull/56342

File: jax/interpreters/mlir.py [hunk 1]
```diff
@@ -1277,7 +1277,7 @@ def emit_python_callback(platform, callback,
       [aval_to_ir_types(aval) for aval in result_avals])
   result_type = ir.TupleType.get_tuple(result_types)
   call_target_name = ("xla_python_gpu_callback"
-                      if platform == "cuda" else "xla_python_cpu_callback")
+                      if platform == "cuda" or platform == "rocm" else "xla_python_cpu_callback")
```
Comment (sharadmv) [inline jax/interpreters/mlir.py:1280]: Could make this `if platform in {"cuda", "rocm"}`

File: tests/array_interoperability_test.py [hunk 2]
```diff
@@ -100,6 +100,7 @@ def testJaxRoundTrip(self, shape, dtype, take_ownership, gpu):
      for shape in all_shapes
      for dtype in dlpack_dtypes))
   @unittest.skipIf(not tf, "Test requires TensorFlow")
+  @jtu.skip_on_devices("rocm") # relevant dlpack protocol is N/A for ROCm ATM
```
Comment (sharadmv) [inline tests/array_interoperability_test.py:103]: How is this related to the callback change?

File: jax/interpreters/mlir.py [hunk 3]
```diff
@@ -1260,9 +1260,9 @@ def emit_python_callback(platform, callback,
   if platform == "cuda" and jax._src.lib.version < (0, 3, 11):
     raise ValueError(
         "`EmitPythonCallback` on CUDA only supported on jaxlib >= 0.3.11")
```
Comment (sharadmv) [inline jax/interpreters/mlir.py:1262]: Please add ROCM here as well

**Review by sharadmv — CHANGES_REQUESTED** [review]

File: tests/array_interoperability_test.py (hunk 2 above)
Comment (rsanthanam-amd) [reply tests/array_interoperability_test.py:103]: It is not (sorry).  But it is necessary for ROCm.  I didn't really want to create a separate PR for this.

File: jax/interpreters/mlir.py (hunk 1 above)
Comment (rsanthanam-amd) [reply jax/interpreters/mlir.py]: Done

File: jax/interpreters/mlir.py (hunk 3 above)
Comment (rsanthanam-amd) [reply jax/interpreters/mlir.py:1262]: Done

File: tests/array_interoperability_test.py (hunk 2 above)
Comment (sharadmv) [reply tests/array_interoperability_test.py:103]: Okay, in that case could you add a `TODO(sharadmv,phawkins):...` and reference a GH issue?

**Review by sharadmv — APPROVED** [review]
Does this PR need to wait on the corresponding TF one to be merged?

**Comment by rsanthanam-amd** [comment]
> Does this PR need to wait on the corresponding TF one to be merged?

No.  The two can merge independently.

File: tests/array_interoperability_test.py (hunk 2 above)
Comment (rsanthanam-amd) [reply tests/array_interoperability_test.py:103]: Just to clarify, should I create a GH issue to add rocm support for dlpack and reference that issue directly in the code?

Comment (rsanthanam-amd) [reply tests/array_interoperability_test.py]: Done

**Comment by sharadmv** [comment]
Thanks!
